core/src/main/kotlin/net/corda/core/flows/FinalityFlow.kt

```diff
@@ -221,18 +222,40 @@ class FinalityFlow private constructor(val transaction: SignedTransaction,
 
     @Suspendable
     private fun notariseAndRecord(): SignedTransaction {
-        val notarised = if (needsNotarySignature(transaction)) {
-            progressTracker.currentStep = NOTARISING
-            val notarySignatures = subFlow(NotaryFlow.Client(transaction, skipVerification = true))
-            transaction + notarySignatures
-        } else {
-            logger.info("No need to notarise this transaction.")
-            transaction
+        val telemetryId = serviceHub.telemetryService.startSpan("notariseAndRecord", flowLogic = this)
+        try {
+            val notarised = if (needsNotarySignature(transaction)) {
+                progressTracker.currentStep = NOTARISING
+                val notarySignatures = subFlow(NotaryFlow.Client(transaction, skipVerification = true))
+                transaction + notarySignatures
+            } else {
+                logger.info("No need to notarise this transaction.")
+                transaction
+            }
+            val recordingTelemetryId = serviceHub.telemetryService.startSpan("recordTransactions", flowLogic = this)
+            try {
+                logger.info("Recording transaction locally.")
+                serviceHub.recordTransactions(statesToRecord, listOf(notarised))
+                logger.info("Recorded transaction locally successfully.")
+                return notarised
+            }
+            catch (ex: Throwable) {
+                recordingTelemetryId.setStatus(StatusCode.ERROR, ex.message ?: "Recording transaction locally failed")
+                recordingTelemetryId.recordException(ex)
+                throw ex
+            }
```

Comment:
This pattern seems to be being used repeatedly throughout many files, adding a lot of repetitive boilerplate code. I wonder if it would make sense to instead provide a function which takes a block of code to execute in the telemetry service. In this case, something like:

```suggestion
            serviceHub.telemetryService.span("recordTransactions", flowLogic = this) {
                logger.info("Recording transaction locally.")
                serviceHub.recordTransactions(statesToRecord, listOf(notarised))
                logger.info("Recorded transaction locally successfully.")
            }
            return notarised
```

Reply:
Done. Yeah, I agree, had originally thought about this and tried initially to use a normal lambda but quasar was reporting un-instrumented methods when suspending. However I've now used inline lambda which works fine.